compiler/rustc_expand/src/proc_macro_server.rs

```diff
@@ -49,158 +49,168 @@ impl ToInternal<token::Delimiter> for Delimiter {
-            Tilde => op!('~'),
-            BinOp(Plus) => op!('+'),
-            BinOp(Minus) => op!('-'),
-            BinOp(Star) => op!('*'),
-            BinOp(Slash) => op!('/'),
-            BinOp(Percent) => op!('%'),
-            BinOp(Caret) => op!('^'),
-            BinOp(And) => op!('&'),
-            BinOp(Or) => op!('|'),
-            BinOp(Shl) => op!('<', '<'),
-            BinOp(Shr) => op!('>', '>'),
-            BinOpEq(Plus) => op!('+', '='),
-            BinOpEq(Minus) => op!('-', '='),
-            BinOpEq(Star) => op!('*', '='),
-            BinOpEq(Slash) => op!('/', '='),
-            BinOpEq(Percent) => op!('%', '='),
-            BinOpEq(Caret) => op!('^', '='),
-            BinOpEq(And) => op!('&', '='),
-            BinOpEq(Or) => op!('|', '='),
-            BinOpEq(Shl) => op!('<', '<', '='),
-            BinOpEq(Shr) => op!('>', '>', '='),
-            At => op!('@'),
-            Dot => op!('.'),
-            DotDot => op!('.', '.'),
-            DotDotDot => op!('.', '.', '.'),
-            DotDotEq => op!('.', '.', '='),
-            Comma => op!(','),
-            Semi => op!(';'),
-            Colon => op!(':'),
-            ModSep => op!(':', ':'),
-            RArrow => op!('-', '>'),
-            LArrow => op!('<', '-'),
-            FatArrow => op!('=', '>'),
-            Pound => op!('#'),
-            Dollar => op!('$'),
-            Question => op!('?'),
-            SingleQuote => op!('\''),
-
-            Ident(name, false) if name == kw::DollarCrate => tt!(Ident::dollar_crate()),
-            Ident(name, is_raw) => tt!(Ident::new(rustc.sess(), name, is_raw)),
-            Lifetime(name) => {
-                let ident = symbol::Ident::new(name, span).without_first_quote();
-                stack.push(tt!(Ident::new(rustc.sess(), ident.name, false)));
-                tt!(Punct::new('\'', true))
-            }
-            Literal(lit) => tt!(Literal { lit }),
-            DocComment(_, attr_style, data) => {
-                let mut escaped = String::new();
-                for ch in data.as_str().chars() {
-                    escaped.extend(ch.escape_debug());
-                }
-                let stream = [
-                    Ident(sym::doc, false),
-                    Eq,
-                    TokenKind::lit(token::Str, Symbol::intern(&escaped), None),
-                ]
-                .into_iter()
-                .map(|kind| tokenstream::TokenTree::token(kind, span))
-                .collect();
-                stack.push(TokenTree::Group(Group {
-                    delimiter: pm::Delimiter::Bracket,
-                    stream,
-                    span: DelimSpan::from_single(span),
-                    flatten: false,
+            let mut op = |s: &str| {
+                let last = s.chars().count() - 1;
+                trees.extend(s.chars().enumerate().map(|(idx, ch)| {
+                    TokenTree::Punct(Punct { ch, joint: joint || idx != last, span })
                 }));
```

Comment:
Possible micro-opt (assuming we don't want to screw around with ASCII-only):
```suggestion
                let mut chars = s.chars();
                let last = chars.next_back().unwrap();
                trees.extend(chars.map(|c| (c, true)).chain([(last, joint)]).map(|(ch, joint)| {
                    TokenTree::Punct(Punct { ch, joint, span })
                }));
```
(this is one of those things where I wish we had `for<const N: usize> |chars: [char; N]|` closures, but alas. though a fun version of this would involve bringing back the macro and having it invoke `push_op::<$lit>(&mut trees, span)`, i.e. instantiating a const-generic function for every possible string literal. overkill, right?)

**EDIT**: I just checked and the `chars` iterator is pretty bad, size-hint-wise (since it divides by "4 bytes per character", rounding up), so asserting ASCII-only might be the way to go. Probably doesn't matter much though.

Reply:
I'll switch it over to using ascii-only `u8` throughout. It'll improve `size_of::<Punct>()` in the client, and won't be too tricky.